Implement NVTXW export

evanramos-nvidia opened the pull request:
This PR contains the second iteration of our NVTXW export feature. The new code looks significantly different than before, keeping in mind the feedback you've given so far. We may still revise some details of these changes, particularly where the Rust code is perhaps too C-like, but overall it is ready for review. It is at a point where getting the code reviewed by someone with more Rust experience will help inform us how we want to finalize it.

I've marked this PR as a draft while we determine where we will put the crate. For review and CI purposes, I've set it up in the subdir `crate-tmp`, but this is temporary. If you have any feedback to offer on the crate in addition to the Legion Prof Viewer code itself, we would definitely appreciate that, since we are new to Rust and still learning.

elliottslaughter commented on src/nvtxw.rs:
```diff
@@ -0,0 +1,385 @@
+        for (_, i) in entries {
+            let c_name = CString::new(i.title).expect("CString::new failed");
+            let events = [legion_nvtxw_event {
+                time_start: (i.time_start.0 as u64).wrapping_add_signed(zero_time),
+                time_stop: (i.time_stop.0 as u64).wrapping_add_signed(zero_time),
```
Are you sure you want a wrapping add here? While it seems unlikely, if (somehow) we actually hit this case, the resulting times will presumably be totally bogus.... It seems better to me to use checked math:

```
i.time_start.0.try_into().unwrap().checked_add(zero_time)
```

Not pretty but it ensures you throw an error if something unexpected happens.

evanramos-nvidia replied:
Done

elliottslaughter reviewed (commented):
I haven't had time to look at the NVTXW code yet, but here are my preliminary thoughts on the part of this that integrates into the prof-viewer code.

Overall this is dramatically better, thanks for doing this work. I appreciate that you work with the existing abstractions rather than working around them. There are a handful of places where I think it would be better to avoid so much hard-coding, but this is hopefully a relatively minor change from what you've got now.

There is one structural change you might consider, which I left a comment about in the source code. At the moment, you essentially read the entire data set into memory, interleave it into one massive table, and then de-interleave it as you write it back out. First of all this uses memory proportional to the size of the profile. Second, it just seems like a lot of work to do this (and probably inefficient). I think that an approach in which you match tiles with their corresponding meta tile, and then write that as a unit, would probably be more efficient. As a bonus it would avoid needing to have so many streams open on the NVTXW side. I think you can do this with fixed memory usage while still allowing a degree of parallelism by configuring the number of outstanding calls you allow.

elliottslaughter commented on crate-tmp/src/nvtxw.rs:
```diff
@@ -0,0 +1,280 @@
+pub const NVTXW3_STREAM_ORDERING_TYPE_UNKNOWN: i16 = 0;
+
+pub fn initialize_simple() -> Result<InterfaceHandle, ResultCode> {
+    let so_name = "libNvtxwBackend.so";
```
Is this going to load an `so` dynamically? Is that it?

evanramos-nvidia replied:
Yes, that is correct.

elliottslaughter commented on crate-tmp/src/nvtxw.rs:
```diff
@@ -0,0 +1,280 @@
+        let s_config = format!("ReportMerge={}", s_merge);
+
+        c_config = CString::new(s_config).expect("CString::new failed");
+        cptr_config = c_config.as_ptr();
```
This code style is a bit risky because it's easy to introduce a dangling pointer if the scope of `c_config` later changes. A less risky coding style would be to store the `CString` in a variable as you do now and then when you use it later:

```
configString: c_config.as_ptr()
```

Thus there's no opportunity for that to dangle.

evanramos-nvidia replied:
Fixed

elliottslaughter commented on crate-tmp/src/nvtxw.rs:
```diff
@@ -0,0 +1,280 @@
+        );
+    }
+
+    if result != NVTXW3_RESULT_SUCCESS {
+        return Err(result);
+    }
```
Bonus points for marking the `nvtxwInitialize` result as `must_use`: https://doc.rust-lang.org/std/hint/fn.must_use.html (and so on for other APIs).

evanramos-nvidia replied:
Thanks for the suggestion, the code is a lot cleaner this way.

For now, I'll hold off on `must_use` because it is still experimental. Once it is stable I can add it.

evanramos-nvidia commented on src/nvtxw.rs:
```diff
@@ -0,0 +1,385 @@
+            }
+        }
+
+        let mut entries: HashMap<u64, EntryContents> = HashMap::new();
```
Done

evanramos-nvidia commented on src/nvtxw.rs:
```diff
@@ -0,0 +1,385 @@
+            }
+        }
+
+        let mut panels: Vec<Vec<Vec<(String, String)>>> = Vec::new();
```
Thanks for the pointer, I've successfully eliminated this structure and worked the details into `walk_entry_list`.

evanramos-nvidia commented on src/nvtxw.rs:
```diff
@@ -0,0 +1,385 @@
+
+                        let key = item_uid.0;
+
+                        match entries.get_mut(&key) {
```
Done

evanramos-nvidia commented on src/nvtxw.rs:
```diff
@@ -0,0 +1,385 @@
+            },
+            nvtxw::PayloadSchemaEntry {
+                flags: 0,
+                type_: nvtxw::NVTX_PAYLOAD_ENTRY_TYPE_NVTX_COLOR,
```
We agree, and will be changing the name of this token.

evanramos-nvidia commented on crate-tmp/src/nvtxw.rs:
```diff
@@ -0,0 +1,280 @@
+        let s_output = os_output
+            .into_string()
+            .expect("OsString::into_string failed");
+        let session_name = match s_output.strip_suffix(".nsys-rep") {
```
Fixed

evanramos-nvidia commented on src/nvtxw.rs:
```diff
@@ -0,0 +1,385 @@
+                    }
+                }
+            }
+            for tile in self.data_source.get_slot_meta_tiles() {
```
So far, I've struggled with implementing batching. I have the code restructured so that I can follow the paradigm used in archive_data.rs, but I think the logic I'm using to match is not correct. In the current push, I've kept the matching logic the same as before, as well as waiting on all tiles rather than batching, so that the output is the also same as before, but I think even that output is not correct, only containing about 2500 events for the circuit example rather than 7500, because the same tiles are matched more than once.

Do you see what I'm doing wrong? I suppose checking EntryUID is not enough--perhaps I should add tile ID comparisons too. I'm also not making use of any ordering so far.

elliottslaughter replied:
I think what you're missing is that there can be multiple `Item`s associated with an `ItemUID`. This happens when a task blocks and goes into a waiting state. So for example, you might have:

 1. From timestamp T0 to T1: task is running
 2. At T1: task blocks
 3. From T1 to T2: task is blocked
 4. At T2: task becomes ready
 5. From T2 to T3: task is ready but not running
 6. At T3: task starts running
 7. From T3 to T4: task is running

In this example you would end up with 4 `Item`s:

 1. `Item` from T0 to T1 in running state
 2. `Item` from T1 to T2 in blocked state
 3. `Item` from T2 to T3 in ready state
 4. `Item` from T3 to T4 in running state
 
 Because they're all the same task, they have the same `ItemUID`.
 
 Here are the invariants I think you can rely on, as long as you always ask for one tile spanning the entire profile with `full=true`. All of the `Item`s with a given `ItemUID` will be guaranteed to be:

 1. On the same row (what you call `item_group` in your code)
 2. Consecutive in array order (i.e., no other items will be interleaved with them)
 3. Consecutive in time (i.e., the stop of one item will always be identical to the start of the next)
 4. The `Item`s and `ItemMeta`s will always appear in the same order and positions in their respective tiles

Let me know if you have any questions.

elliottslaughter replied:
By the way, some of the invariants above would have been violated by previous Legion versions, but if you update to the newest `master` they should all be preserved now.

elliottslaughter replied:
While we're here, I wanted to remind you about this comment I made last time:

https://github.com/StanfordLegion/prof-viewer/pull/54#pullrequestreview-2006506214

Fundamentally, you hit the issue with `ItemUID`s because you're slurping the entire profile into memory and matching `Item`s across tiles. There are two issues. The first is with managing the invariants of multiple items as you noted. The second is that it uses memory proportional to the size of the profile.

I've seen cases where this simply won't fit into memory. Mike Bauer showed me one the other day from cuNumeric. So this is simply a no-go for use cases that I'm pretty sure you'll care about. But since it should also simplify the implementation to do things are tile granularity, I think it would just make the code nicer to deal with overall.

evanramos-nvidia replied:
I believe I have finally figured out how to implement the batching you've requested. Please let me know if the code is indeed doing what you intended.

elliottslaughter replied:
Structurally, I think this should work. I provided a couple of suggestions in the code: one should simply the structure further, and the other is a code style issue that would make the code more straightforward.

I have not re-reviewed the NVTXW code specifically. Would you like me to do so now?

evanramos-nvidia replied:
Thank you for your review, I see how using a pair of `Option`s could simplify the code further.

> I have not re-reviewed the NVTXW code specifically. Would you like me to do so now?

Yes please, the NVTXW crate code is ready for further review.

evanramos-nvidia commented:
Updated to fix introduced clippy diagnostics

elliottslaughter commented:
I'm on vacation but I will hit the "run" button as long as you keep pushing changes.

I'll get back to you with a review on August 26.

elliottslaughter commented on src/nvtxw.rs:
```diff
@@ -0,0 +1,427 @@
+}
+
+type ResultVec = Vec<(EntryID, String, String)>;
+type UnmatchedTileHold = BTreeMap<EntryID, SlotTile>;
+type UnmatchedMetaTileHold = BTreeMap<EntryID, SlotMetaTile>;
```
Instead of two maps I think you could make this a `BTreeMap<EntryID, (Option<SlotTile>, Option<SlotMetaTile>)>`. At that point you can do a single pass with [`retain`](https://doc.rust-lang.org/std/collections/struct.BTreeMap.html#method.retain) and trigger only on the entries that contain `(Some(_), Some(_))` (i.e., both sides are filled).

evanramos-nvidia replied:
I've made this change to the code. How does this part look now?

elliottslaughter replied:
I added a comment on one place that can be simplified, but otherwise it looks good. Thanks.

evanramos-nvidia replied:
Great!

elliottslaughter commented on src/nvtxw.rs:
```diff
@@ -0,0 +1,427 @@
+
+pub struct NVTXW<T: DeferredDataSource> {
+    data_source: CountingDeferredDataSource<T>,
+    output: Option<PathBuf>,
+    merge: Option<PathBuf>,
```
See my note here and apply the same thing:

https://gitlab.com/StanfordLegion/legion/-/merge_requests/1001#note_2076285077

In general, it's considered a best practice not to pass around concrete path types, as `AsRef<Path>` gives you more flexibility in passing around e.g. `OsString` without requiring any conversion.

evanramos-nvidia replied:
I've tried implementing this, but am getting errors.

```rs

pub fn session_begin_simple<P1, P2>(
    interface: InterfaceHandle,
    output: Option<P1>,
    merge: Option<P2>,
) -> Result<SessionHandle, ResultCode>
where
    P1: AsRef<Path>,
    P2: AsRef<Path>,
{
    let session_begin = interface
        .SessionBegin
        .ok_or(NVTXW3_RESULT_INVALID_ARGUMENT)?;

    let c_output = output
        .as_ref()
        .and_then(|x| {
            let p = Path::new(x);
            // ...
```

```
146 |             let p = Path::new(x);
    |                     --------- ^ the trait `AsRef<OsStr>` is not implemented for `P1`
    |                     |
    |                     required by a bound introduced by this call
    |
```

elliottslaughter replied:
You'd need to do something like: 

```rust
let output = output.map(AsRef::as_ref);
let merge = merge.map(AsRef::as_ref);
```

And then you'll have them both as `Option<Path>` for the rest of the function.

evanramos-nvidia replied:
```
error[E0631]: type mismatch in function arguments
   --> src/nvtxw.rs:144:29
    |
144 |     let output = output.map(AsRef::as_ref);
    |                         --- ^^^^^^^^^^^^^
    |                         |   |
    |                         |   expected due to this
    |                         |   found signature defined here
    |                         required by a bound introduced by this call
    |
    = note: expected function signature `fn(P1) -> _`
               found function signature `fn(&_) -> _`
note: required by a bound in `Option::<T>::map`
   --> /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/core/src/option.rs:1070:5
help: consider wrapping the function in a closure
    |
144 |     let output = output.map(|arg0: P1| AsRef::as_ref(&arg0));
    |                             ++++++++++              +++++++

```

evanramos-nvidia replied:
I've opted for making these `OsString`s instead.

elliottslaughter commented on crate-tmp/src/nvtxw.rs:
```diff
@@ -0,0 +1,248 @@
+    Ok(())
+}
+
+pub fn initialize_simple() -> Result<InterfaceHandle, ResultCode> {
```
Is there a rationale on the `_simple` suffix of a lot of these API calls?

evanramos-nvidia replied:
My intent with the `_simple` APIs is to use sensible defaults for some of the configurable properties the C APIs have, such as `SessionAttributes`, `StreamAttributes`, and `so_name`, so that the Rust user code doesn't have to bother with them.

elliottslaughter commented on crate-tmp/src/nvtxw.rs:
```diff
@@ -0,0 +1,248 @@
+        .ok_or(NVTXW3_RESULT_INVALID_ARGUMENT)?;
+
+    let mut c_output: CString = CString::new("").expect("CString::new failed");
+    if let Some(os_output) = o_output {
```
Ok, there's a bunch of stuff to say about this:

 1. This code double-allocates the `CString` in the case where `o_output` is `None`
 2. The code uses an empty `CString` as an error case which is weird and non-idiomatic
 3. The logic itself still uses direct string manipulation
 4. The line `CString::new(...).expect(...)` gets repeated all over the place

If I understand what you're trying to do, you'd get a very similar logic with:

```rust
let c_output = o_output.and_then(|o| Path::new(o).file_stem()).map(|s| CString::new(s).expect("output contains nul bytes"));
```

Note that this strips *any* file extension rather than just `.nsys-rep`. If you feel very strongly about that specific behavior (and that if a user who uses a non-default extension should *not* have it stripped), then we'd have to tweak the approach.

Since you use the empty case as an error anyway, it doesn't seem worth allocating a `CString` just to check its emptiness. Just let it be `None` on failure and then later on you can do:

```rust
c_output.map_or_else(null_mut, |o| o.as_ptr())
```

As a general principle, you do not want to use sentinels in Rust, and instead you want to directly encode error states so that when things go bad it's well-defined what happens.

evanramos-nvidia replied:
I've updated this part of the code. (It turns out `.as_ref()` is important in multiple places.) Please let me know if it can be improved further.

elliottslaughter commented on crate-tmp/src/nvtxw.rs:
```diff
@@ -0,0 +1,248 @@
+    stream_name: String,
+    domain_name: String,
+) -> Result<StreamHandle, ResultCode> {
+    let func = interface.StreamOpen.ok_or(NVTXW3_RESULT_INVALID_ARGUMENT)?;
```
If it were me I'd name this something more descriptive like `stream_open`, so that when you go call it down below it's obvious what you're calling. I get that this follows a pattern and some of the later API calls are shorter (thus making it more obvious), but in these longer API calls, you have to read back a bit to figure out what is being called.

evanramos-nvidia replied:
Done

elliottslaughter commented on crate-tmp/src/nvtxw.rs:
```diff
@@ -0,0 +1,248 @@
+
+pub fn session_begin_simple(
+    interface: InterfaceHandle,
+    o_output: Option<PathBuf>,
+    o_merge: Option<PathBuf>,
```
Generally speaking it's not typical to see Rust code encoding the types into the names of variables. Usually if you are going through some sort of a multi-step type conversion process, you chain the conversion operators like I suggest below.

evanramos-nvidia replied:
Fixed

elliottslaughter commented on src/nvtxw.rs:
```diff
@@ -0,0 +1,427 @@
+                let payloads = [
+                    nvtxw::PayloadData {
+                        schemaId: LEGION_NVTXW_PAYLOAD_NAME_SCHEMA_ID,
+                        size: usize::MAX,
```
Is this supposed to be `size_of::<usize>()`?

evanramos-nvidia replied:
`usize::MAX` functions correctly here. In the C API, it would be `SIZE_MAX`, `(size_t)-1`, or `~(size_t)0`, etc. The NVTX extended payload spec uses this value as a sentinel to mean the backend should call `strlen` on the C string passed in.

elliottslaughter commented on crate-tmp/src/nvtxw.rs:
```diff
@@ -0,0 +1,248 @@
+pub fn event_write(
+    interface: InterfaceHandle,
+    stream: StreamHandle,
+    events: &[PayloadData],
```
I'm sort of torn about this. On the one hand, I can see you're doing this to avoid copying data. On the other hand, it exposes the guts of `PayloadData` and the need to properly encode all the right C data structures to the user. This makes for fairly non-idiomatic Rust code on the end-user side.

I guess it's fine for now, but it seems like a prime candidate for making a better abstraction down the line.

evanramos-nvidia replied:
I agree that the resulting user code is not ideal. My understanding is that there is room for improvement by using Rust's macro system to populate some details of our structures, as long as your structure already has a C representation.

elliottslaughter reviewed (commented):
I reviewed the Rust NVTXW crate.

Overall my highest level comment is that you may want to think about exactly how high- or low-level you want this interface to be. The current interface is serviceable and does, strictly speaking, remove `unsafe` from user code. But the code that users have to write is still a bit questionable: ideally, users wouldn't have to think that hard about struct packing, `CString`, and raw pointers at all. So this may be fine for now, but just recognize that as written it's not a great example of best-in-class Rust API wrappers for C.

I added some minor code comments to improve quality in various ways throughout the code, and suggested some more idiomatic practices.

elliottslaughter commented:
Some other things you might think about for NVTXW, which I did not specifically address:

  * Can you add tests for NVTXW? Rust has a pretty good test infrastructure, and you can write both unit and integration tests in the crate (which will both be run by `cargo test`). Here's the documentation on how to do this: https://doc.rust-lang.org/book/ch11-01-writing-tests.html
  * Once you have tests it's considered a best practice to run your code through Miri, which is sort of like asan/ubsan for Rust. This matters because of the unsafe code in the crate; otherwise Rust's type system would make it unnecessary to do anything of the sort: https://github.com/rust-lang/miri
  * If you were to deploy this in a standalone repo, you would want to copy the test infrastructure inside `.github/workflows/rust.yml` (or implement something like it). It has a bunch of standard CI jobs to check the build, tests, format, lints, etc.

evanramos-nvidia commented on crate-tmp/Cargo.toml:
```diff
@@ -0,0 +1,13 @@
+
+[build-dependencies]
+bindgen = ">=0"
+cc = ">=1"
```
Done

evanramos-nvidia commented on src/nvtxw.rs:
```diff
@@ -0,0 +1,427 @@
+
+            let mut remove_these: Vec<EntryID> = Vec::new();
+
+            unmatched_tiles.iter().for_each(|(entry_id, tile)| {
```
These instances of `for_each` are now gone. Should I adjust the two remaining instances in this file, which operatie on `zip(...)`?

elliottslaughter replied:
Yes, if it were me, I'd rewrite

```rust
zip(X, Y).for_each(|(x, y)| {
```

as

```rust
for (x, y) in zip(X, Y) {
```

evanramos-nvidia replied:
Done

evanramos-nvidia commented:
Thanks for your feedback, it's really helpful both to make the code better and to build my Rust skills.

I've updated the branch to address some items, and I will continue with the rest and update it further. Right now, I am particularly interested in feedback on my revised implementation of tile matching, using a single `BTreeMap` where the value is a pair of `Option`s.

elliottslaughter commented on src/nvtxw.rs:
```diff
@@ -0,0 +1,419 @@
+            let slot_meta_tiles = data_source.get_slot_meta_tiles();
+
+            for tile in slot_tiles {
+                match unmatched_tiles.entry(tile.entry_id.clone()) {
+                    Entry::Occupied(mut entry) => {
+                        let (ut, _um) = entry.get_mut();
+                        *ut = Some(tile);
+                    }
+                    Entry::Vacant(entry) => {
+                        entry.insert((Some(tile), None));
+                    }
+                }
```
I think you could shorten this to:

```rust
unmatched_tiles.entry(tile.entry_id.clone()).or_insert((None, None)).0 = Some(tile);
```

That is, the `or_insert` ensures we get a default value if it's missing, but otherwise we're just going to overwrite it anyway, so no need to branch.

And similarly for the meta case but make sure to access `.1` (i.e., the second field in the tuple).

evanramos-nvidia replied:
Done. I'm really glad there is a simpler version of this.

evanramos-nvidia commented:
The CI is now failing with

```
Clippy: nvtxw_bindings.rs#L3
use of unstable library feature 'offset_of'
```

The only thing I can think of that might have introduced this is specifying the version of bindgen in the crate's Cargo.toml.

elliottslaughter commented:
Try reverting to bindgen `0.69`? That seems to be what you used in the last successful build:

https://github.com/StanfordLegion/prof-viewer/actions/runs/10411469581/job/28835561058

elliottslaughter commented on crate-tmp/src/nvtxw.rs:
```diff
@@ -0,0 +1,246 @@
+        .SessionBegin
+        .ok_or(NVTXW3_RESULT_INVALID_ARGUMENT)?;
+
+    let c_output = Some(PathBuf::from(output))
```
If `output` is not an `Option` then I don't think we need all of this logic to re-wrap it?

evanramos-nvidia replied:
I kept this to maintain the clean approach that the chained `map`/`and_then` calls provide. Otherwise I was looking at making `c_output` into a mutable variable. I can drop the chaining.

evanramos-nvidia replied:
I've updated this part of the code.

elliottslaughter commented on crate-tmp/src/nvtxw.rs:
```diff
@@ -0,0 +1,246 @@
+        .and_then(|s| Some(CString::new(s.to_str()?).expect("c_output CString::new failed")));
+
+    let c_config = merge
+        .map(PathBuf::from)
+        .map(|p| format!("ReportMerge={}", p.display()))
```
I'm not entirely sure why we're doing the `PathBuf::from` given that we then turn around and immediately pass it to `format!`. Could we do:

```suggestion
        .map(|p| format!("ReportMerge={}", p.as_ref()))
```

evanramos-nvidia replied:
Done

elliottslaughter commented on src/nvtxw.rs:
```diff
@@ -0,0 +1,403 @@
+        let c_field_name_name = CString::new("name").expect("CString::new failed");
+        let c_field_name_color = CString::new("color").expect("CString::new failed");
+
+        // The C string containing the event's name must be provided in a special way.
```
Let's either explain what the "special way" is, or else remove this comment if it's not necessary. As it is this is too vague to be helpful.

evanramos-nvidia replied:
I changed the comment.

elliottslaughter commented on Cargo.toml:
```diff
@@ -17,7 +17,8 @@ rust-version = "1.74"
 default = []
 client = ["dep:reqwest", "dep:url"]
-server = ["dep:actix-cors", "dep:actix-web"]
+server = ["dep:time", "dep:actix-cors", "dep:actix-web"]
```
I'm not sure why `dep:time` is being added since you're not modifying the `server` component?

evanramos-nvidia replied:
See below

elliottslaughter commented on Cargo.toml:
```diff
@@ -50,9 +51,14 @@ url = { version = "2", optional = true }
 
 
 # server:
+time = { version = "0.3", optional = true }
```
I'm not sure why this is being added under `server`? Are you actually using this at all?

evanramos-nvidia replied:
I added the commit `Work around build failure in time v0.3.34 (<-- cookie <-- actix-web)` to work around this build failure that appeared with Rust 1.80:

```
$ ./check.sh
+ cargo check --workspace --no-default-features --all-targets
    Checking legion_prof_viewer v0.3.0 (/home/evanr/sbf/legion-prof-viewer)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.50s
+ cargo check --workspace --no-default-features --features client --all-targets
    Checking legion_prof_viewer v0.3.0 (/home/evanr/sbf/legion-prof-viewer)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.56s
+ cargo check --workspace --no-default-features --features server --all-targets
   Compiling time-macros v0.2.17
    Checking time v0.3.34
error[E0282]: type annotations needed for `Box<_>`
  --> /home/evanr/.cargo/registry/src/index.crates.io-6f17d22bba15001f/time-0.3.34/src/format_description/parse/mod.rs:83:9
   |
83 |     let items = format_items
   |         ^^^^^
...
86 |     Ok(items.into())
   |              ---- type must be known at this point
   |
help: consider giving `items` an explicit type, where the placeholders `_` are specified
   |
83 |     let items: Box<_> = format_items
   |              ++++++++

For more information about this error, try `rustc --explain E0282`.
error: could not compile `time` (lib) due to 1 previous error
```

Rust 1.81 says this:

```
error[E0282]: type annotations needed for `Box<_>`
  --> /home/evanr/.cargo/registry/src/index.crates.io-6f17d22bba15001f/time-0.3.34/src/format_description/parse/mod.rs:83:9
   |
83 |     let items = format_items
   |         ^^^^^
...
86 |     Ok(items.into())
   |              ---- type must be known at this point
   |
   = note: this is an inference error on crate `time` caused by an API change in Rust 1.80.0; update `time` to version `>=0.3.35` by calling `cargo update`

For more information about this error, try `rustc --explain E0282`.
```

So I ran `cargo update` which updated the `Cargo.lock` file. It also resolved this error. However, it exposed a new one.

```
$ ./check.sh
+ cargo check --workspace --no-default-features --all-targets
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.30s
+ cargo check --workspace --no-default-features --features client --all-targets
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.18s
+ cargo check --workspace --no-default-features --features server --all-targets
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.13s
+ cargo check --workspace --all-features --all-targets
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.14s
+ cargo check --workspace --no-default-features --lib --target wasm32-unknown-unknown
    Checking eframe v0.25.0
error[E0308]: mismatched types
   --> /home/evanr/.cargo/registry/src/index.crates.io-6f17d22bba15001f/eframe-0.25.0/src/web/mod.rs:199:16
    |
199 |         if let Some(clipboard) = window.navigator().clipboard() {
    |                ^^^^^^^^^^^^^^^   ------------------------------ this expression has type `Clipboard`
    |                |
    |                expected `Clipboard`, found `Option<_>`
    |
    = note: expected struct `Clipboard`
                 found enum `std::option::Option<_>`

For more information about this error, try `rustc --explain E0308`.
error: could not compile `eframe` (lib) due to 1 previous error
```

I would definitely like a better way to avoid this problem.

elliottslaughter replied:
I think you can remove this line and do:

```
cargo update time
```

And it will update the contents of the `Cargo.lock` file to use the latest version.

evanramos-nvidia replied:
That seemed to work.

elliottslaughter commented:
@evanramos-nvidia I added some more (mostly minor) comments on the code.

What are your plans in terms of merging this at this point? We have feature freeze for Legion 24.09.0 coming up on 9/16.

evanramos-nvidia commented:
> @evanramos-nvidia I added some more (mostly minor) comments on the code.

Thanks for your continued attention to detail on things.

> What are your plans in terms of merging this at this point? We have feature freeze for Legion 24.09.0 coming up on 9/16.

We would like to merge the feature in time for 24.09.0.

evanramos-nvidia commented on crate-tmp/src/nvtxw.rs:
```diff
@@ -0,0 +1,248 @@
+use crate::nvtxw_bindings::NVTXW3_RESULT_SUCCESS;
+use crate::nvtxw_bindings::NVTXW3_STREAM_ORDERING_SKID_NONE;
+
+pub const NVTX_PAYLOAD_ENTRY_FLAG_POINTER: u64 =
+    nvtxw_bindings::NVTX_PAYLOAD_ENTRY_FLAG_POINTER as u64;
```
I implemented this change locally but was discouraged to find that Rust has no implicit conversion from an enum to its underlying integer type, and even moreso that there is no way to convert an integer to an enum of corresponding value other than a `match` block listing every possibility. Adjusting the API so that these are no longer an issue would involve using less and less of bindgen's output directly and instead generating it ourselves or maintaining it manually. I am still investigating a suitable path forward.

elliottslaughter replied:
There are several third-party crates to do this, but in the main compiler we use [`num_enum`](https://docs.rs/num_enum/latest/num_enum/).

evanramos-nvidia replied:
I've revisited this change with the addition of the `num_enum` crate, which I found to be helpful. Unfortunately, the resulting code is still a bit ugly, because every use of the flag/type enums in the user code must be accompanied by `.into()` to fit into the struct types generated by bindgen, and also because the BitOr trait cannot be applied to enums. Additionally, I encountered this confusing error at the first instance of a flag with `.into()` in the user code, despite the `flags` member and the representation of the `PayloadEntryFlag` enum being `u64`.

```
error[E0283]: type annotations needed
   --> src/nvtxw.rs:248:58
    |
248 |             flags: nvtxw::PayloadEntryFlag::EventMessage.into()
    |                                                          ^^^^
    |
    = note: cannot satisfy `_: std::convert::From<PayloadEntryFlag>`
    = note: required for `PayloadEntryFlag` to implement `Into<_>`
help: try using a fully qualified path to specify the expected types
    |
248 |             flags: <PayloadEntryFlag as Into<T>>::into(nvtxw::PayloadEntryFlag::EventMessage)
    |                    ++++++++++++++++++++++++++++++++++++                                     ~

```

At this point I'm wondering what I should do now. I suspect that the difficulty I'm encountering is because using enums in this way is not very Rust-like. For example, instead of bitwise ORing together flags, a more idiomatic approach might look something more like `std::fs::OpenOptions`. That said, redesigning the API in time for merge is an imposing proposition.

Another idea I had that might offer some degree of improvement is something like this:

```rs
#[allow(non_snake_case)]
pub mod PayloadEntryFlag {
    use crate::nvtxw_bindings;
    pub const POINTER: u64 = nvtxw_bindings::NVTX_PAYLOAD_ENTRY_FLAG_POINTER as u64;
    pub const EVENT_MESSAGE: u64 = nvtxw_bindings::NVTX_PAYLOAD_ENTRY_FLAG_EVENT_MESSAGE as u64;
    // ...
}
```

```rs
        let name_schema = [nvtxw::PayloadSchemaEntry {
            flags: nvtxw::PayloadEntryFlag::EVENT_MESSAGE
                | nvtxw::PayloadEntryFlag::ARRAY_ZERO_TERMINATED,
            // ...
```

elliottslaughter replied:
Let's defer this item for now. I don't want to rush a last-minute API change.

Longer-term, I agree that `std::fs::OpenOptions` is a good example to learn from. We'll still need to play with it to see how much it makes sense. But if you could write an API like, e.g.:

```rust
nvtxw::PayloadEntryFlag::event().array_zero_terminated()
```

Maybe that's better. Certainly, it provides an opportunity for validation which bitwise-or-ing of flags does not provide, and the builder pattern is more Rust-ish in general. The main thing is just the sheer number of flags involved. But maybe that's ok.

evanramos-nvidia replied:
Thanks for your flexibility. I definitely want to do my research in order to design a robust API.

elliottslaughter commented:
If you merge and/or rebase the latest `master`, you will see a new `Check macOS` job that runs on Mac. That will allow you to see the Mac failures even if you do not personally develop on Mac. (Having said that, developing via CI is a massive pain in the rear, so you may still want to get yourself a Mac to build on.)

We do need to keep the Mac build working because many of our users are on Mac.

elliottslaughter commented on crate-tmp/src/nvtxw.rs:
```diff
@@ -0,0 +1,248 @@
+    let s_output: &OsStr;
+
+    // avoid "*.nsys-rep.nsys-rep"
+    let path_output = PathBuf::from(output);
+    if let Some(ext) = path_output.extension() {
+        if ext == "nsys-rep" {
+            s_output = path_output.file_stem().unwrap();
+        } else {
+            s_output = path_output.as_os_str();
+        }
+    } else {
+        s_output = path_output.as_os_str();
+    }
```
How about:

```suggestion
    let path_output = PathBuf::from(output);
    path_output.extension().and_then(
        |ext| {
            if ext == "nsys-rep" {
                path_output.file_stem()
            } else {
                None
            }
        }).unwrap_or(path_output.as_os_str());
```

It's not that much shorter, but it collapses the two else cases and keeps the nice functional structure you had before, but without extra wrapping of the `Some(...)` at the beginning.

elliottslaughter replied:
Sorry, missed the variable binding. Let me do this over.

elliottslaughter commented on crate-tmp/src/nvtxw.rs:
```diff
@@ -0,0 +1,248 @@
+        .SessionBegin
+        .ok_or(NVTXW3_RESULT_INVALID_ARGUMENT)?;
+
+    let s_output: &OsStr;
+
+    // avoid "*.nsys-rep.nsys-rep"
+    let path_output = PathBuf::from(output);
+    if let Some(ext) = path_output.extension() {
+        if ext == "nsys-rep" {
+            s_output = path_output.file_stem().unwrap();
+        } else {
+            s_output = path_output.as_os_str();
+        }
+    } else {
+        s_output = path_output.as_os_str();
+    }
```
Ok, trying this again, now with `s_output` properly set:

```suggestion
    // avoid "*.nsys-rep.nsys-rep"
    let path_output = PathBuf::from(output);
    let s_output = path_output.extension().and_then(
        |ext| {
            if ext == "nsys-rep" {
                path_output.file_stem()
            } else {
                None
            }
        }).unwrap_or(path_output.as_os_str());
```

As in my last attempt, it collapses the two else clauses and is generally more functional, doesn't involve setting variables, and tries to do everything in exactly one place.

elliottslaughter replied:
I've been thinking more about this and I think this is still to clever. I think the effect you want can be accomplished more clearly with:

```rust
let mut path_output = PathBuf::from(output);
if path_output.extention() == Some("nsys-rep") {
  path_output.set_extention("");
}
let s_output = path_output.as_os_str();
```

Exactly one conditional, clearly stated. You get a mutable variable but I think it's pretty clear what's going on (clear the extention when it is "nsys-rep"). `s_output` follows strictly after and is computed very straightforwardly.

evanramos-nvidia replied:
I like this too. Applied.

elliottslaughter commented on src/nvtxw.rs:
```diff
@@ -0,0 +1,394 @@
+        num_requests: u64,
+    ) {
+        while data_source.outstanding_requests() > num_requests {
+            let _summary_tiles = data_source.get_summary_tiles();
```
I'm not sure we need this since we don't actually use summary tiles? I think it's ok to delete this outright.

evanramos-nvidia replied:
I turned this back into a comment and added another comment before it.

elliottslaughter commented:
In order to get this into the release, here's what we need to do:

 * Fix the macOS build
 * Address any outstanding comments (I think we're getting there, just a few more)
     * Note: make sure you see https://gitlab.com/StanfordLegion/legion/-/merge_requests/1001#note_2097198025 as that may require work here
 * Split the nvtxw crate back out into its own repository and publish it
 * Update this PR to reference that (either via `git` URL or version, depending on what you chose for the last step)
 * Merge this PR
 * I'll do a release here (this takes less than an hour)
 * Update https://gitlab.com/StanfordLegion/legion/-/merge_requests/1001 to refer to the new version and address any remaining comments on that MR
 * Merge that MR

So we have a bit of work still to do but it's possible that we'll get there if we make steady progress.

evanramos-nvidia commented:
@elliottslaughter I've fixed the build on macOS. Thanks for bringing that to my attention.

elliottslaughter commented on crate-tmp/src/nvtxw.rs:
```diff
@@ -0,0 +1,241 @@
+        path_output.set_extension("");
+    }
+    let s_output = path_output
+        .as_os_str()
+        .to_str()
```
I think you can do `.to_str()` on a `PathBuf` directly so you can skip the indirection through `.as_os_str()`:

```suggestion
        .to_str()
```

evanramos-nvidia replied:
Fixed

evanramos-nvidia commented:
>   * Note: make sure you see https://gitlab.com/StanfordLegion/legion/-/merge_requests/1001#note_2097198025 as that may require work here

I've implemented this request.

elliottslaughter commented on crate-tmp/src/nvtxw.rs:
```diff
@@ -0,0 +1,281 @@
+        path_output.set_extension("");
+    }
+
+    // however, we are responsible for when the output already exists
+    let mut path_test = path_output.with_extension(extension);
+    if path_test.is_dir() {
+        return Err(NVTXW3_RESULT_INVALID_ARGUMENT);
+    } else if path_test.exists() {
+        if force {
+            println!("Overwriting {:?}", &path_test);
+        } else {
+            let mut i = 1;
+            let retry_limit = 100;
+            loop {
+                let e = format!("{}.{}", i, extension);
+                path_test = path_output.with_extension(e);
+                if !path_test.exists() {
+                    path_output.set_extension(format!("{}", i));
+                    break;
+                } else if i >= retry_limit {
+                    return Err(NVTXW3_RESULT_FAILED);
+                }
+                i += 1;
+            }
+        }
+    }
```
Unfortunately this code is going to be racy. The equivalent code here is written fairly carefully to make exactly one filesystem API call per attempt:

https://github.com/StanfordLegion/prof-viewer/blob/master/src/archive_data.rs#L21-L46

If I understand correctly, NVTXW will overwrite whatever file you create here. So, I think the safe equivalent of this code is to not ever call `.is_dir()` or `.exists()` but to just go and create the file directly. You can do this with the `File::create_new` API:

https://doc.rust-lang.org/std/fs/struct.File.html#method.create_new

This will fail if the file already exists, so it will not touch any existing files. It will also fail if the path is a directory, so I don't think we need any separate checks for that case.

It's always possible for some other process or user to mess with the file system between when we create this file, and when NVTXW attempts to overwrite it. However, if the process is merely creating files without removing them, then I think this avoids the race. (E.g., think of if the user is running multiple copies of the profiler in parallel.)

evanramos-nvidia replied:
Ah, I understand! It's like how the only way to know a lock is available is by actually holding it. I've updated the code to use `File::create_new`.

elliottslaughter replied:
I missed that `File::create_new` was added in Rust 1.77. If I understand correctly, the corresponding `OpenOptions` has existed since 1.9, so you can still write this in Rust 1.74, it's just more obnoxious to implement:

```rust
// TODO: replace with File::create_new after Rust 1.77
fn file_create_new<P: AsRef<Path>>(path: P) -> io::Result<File> {
  OpenOptions::new().read(true).write(true).create_new(true).open(path.as_ref())
}
```

And then call `file_create_new` instead of `File::create_new`.

evanramos-nvidia replied:
Cool, this fixed the CI build.

evanramos-nvidia commented:
@elliottslaughter

> In order to get this into the release, here's what we need to do:

Other than the discussion about the enums in the NVTXW crate, I believe I have addressed all review feedback. Once I have resolved it and have your approval, I am prepared to proceed with the remaining steps for publishing the crate and updating the branches. We should be able to complete this within the week, in time for your merge window.

elliottslaughter commented:
@evanramos-nvidia I wanted to address one point mentioned in your previous comment:

> We should be able to complete this within the week, in time for your merge window.
> https://github.com/StanfordLegion/prof-viewer/pull/54#issuecomment-2354225878

I posted the release timeline here on August 29:
https://legion.zulipchat.com/#narrow/stream/187787-general/topic/legion.20release.2024.2E09.2E0/near/466074958

In particular, feature freeze was September 16.

The intent of our release system is to follow a train model. The goals are (a) to run frequent releases so it matters less what specific release a feature gets into, and (b) merge features when they're ready (rather than sprint to get them ready by some deadline). Even though we have a cutoff date we usually shy away from large merges toward the end of the cycle anyway.

Based on the fact that this is new code being added and we're past the feature freeze date, it would be my plan to merge this right after the release so that it's early in the next cycle. My understanding is that cuNumeric tracks the `master` branch so this will make it available to cuNumeric as soon as they update, so it will not delay adoption on their side.

Let me know if you have any questions.

elliottslaughter commented:
@evanramos-nvidia are you ready to move forward with this? We can make this available to the cuNumeric folks as soon as it hits `master`.

evanramos-nvidia commented:
> * Split the nvtxw crate back out into its own repository and publish it (not necessarily to crates.io, can be just Github for now)
> * Update this PR to reference that (either via `git` URL or version, depending on what you chose for the last step)

Updated to complete these steps.

elliottslaughter commented on Cargo.toml:
```diff
@@ -53,6 +54,8 @@ url = { version = "2", optional = true }
 actix-cors = { version = "0.6", optional = true }
 
+# nvtxw:
+nvtxw = { git = "https://github.com/NVIDIA/NVTX.git", branch = "nvtxw-rs", optional = true }
```
One thing I just wanted to check is whether this is a long-term stable branch. I'm not sure what would happen if we were to merge this here and then you deleted the upstream branch. Might be better to merge the Rust code upstream before we commit this on our end.

tcourtneynv replied:
Yes, it is a long-term stable branch.

evanramos-nvidia replied:
Please hold that thought while we discuss this.

evanramos-nvidia replied:
We are still determining what our final course of action will be, and are trying to arrive at a conclusion ASAP. For the moment, if you could continue to approve CI runs if/when I push an update to the branch, that would be crucially helpful.

evanramos-nvidia replied:
We have decided to publish the nvtxw crate to crates.io as v0.1.0. As of today it is basically ready, pending one last CI verification. Once we publish it I will update this branch to reference the actual crate and remove this PR's draft status.

evanramos-nvidia replied:
Crate published, branch updated, marked ready for review.

elliottslaughter replied:
@evanramos-nvidia @tcourtneynv CI is passing, let me know if you're ready to merge.

evanramos-nvidia replied:
Yes, we are ready for merge.

evanramos-nvidia replied:
Great! I've updated the Legion MR.

evanramos-nvidia commented:
Elliott, thank you for your help in reviewing this PR. It has been valuable to get your feedback and coaching since this is the first Rust project I've worked on.

elliottslaughter commented:
Yup, and thanks for putting in the work to make all the changes. It's been a long road but I think the code is in a much better place now, with a much better foundation for building future work.

The Legion side MR got merged so this should be available to the Legate folks as soon as they update Legion.
